### ISSUE TITLE
Folders and Sub-Vault Feature

### ISSUE THREAD

**Issue by mdowais** [padloc/padloc#533]
# Folders
To allow creation of folders, in which vault items can be created or moved to. This would allow the vault to be organized, rather than having everything on the index page of the vault.

# Sub-Vault
Folders will allow provision to organize, but introducing folder would be more better if it was possible to give ACL for the same. I guess the core revolves around the concept of Vaults.  But mere folders which are meant for organizing, cannot be giving individual crpto stature of a vault. Hence I Propose, Sub-Vault. 

My Idea for Sub-Vault is that, a Sub-Vault is a Vault which is not shown in the Navigation Panel and must be a child inside a Main Vault in a given path, This allows the Sub-Vault to override the ACL permission of the Main Vault. It can be made to allow/deny access to members as it is a separate vault, but will allow full access to owner/admin of vault. A Folder can be promoted to a Sub-Vault which would give it crypto stature and Demoted when necessary.

This would benefit in cases like, 
- sharing a Sub-Vault(folder) inside a vault to someone else, where the folder is converted to a vault.

Note : I really loved the Security behind it, and was quite intrigued. But, I am into C# and Angular, which limited my ability to contribute to this project.

**Comment by mdowais** [comment]
I have been trying to understand the structure, will try to contrib.

**Comment by MaKleSoft** [comment]
Hi @moigarmo! Thanks for your suggestions! This is not the first time folders and sub-vaults have been suggested, and it's not a bad idea in general! However, we've made the conscious decision in the past to not implement any advanced substructures like this for various reasons.

The first and most obvious reason is that it introduces additional complexity in the code base. The addition of subvaults in particular would require significant changes to the overall data structure and has non-trivial implications on the cryptographic design as well.

The second reason is that I personally don't believe that it would actually add that much value to the majority of the user base. One of the things distinguishing Padloc from other product is it's **simplicity**, and with every added feature we necessarily lose some of that simplicity. We already have [tags](https://docs.padloc.app/manual/vaults/#tags) for organising vault items which in many ways they are superior to folders, anyway.

So, long story short... Although we absolutely appreciate your suggestion, I'm afraid this is not something we're willing to consider at this time. Please don't let this discourage you from offering more suggestions in the feature though!

**Comment by mdowais** [comment]
Thanks for the reply, I recognize that introducing Sub-Vault would introduce additional complexity on the code base and the cryptographic design as-well, and as you say, when simplicity at mind, i concur to your view.

But, Tags are great to Tag to *Categorize* but I fail to recognize how it would be beneficial as an *Organizer*.

If the Items were to be inside folder(s), it would look much more simpler and Clean. 
The Current Implementation looks like having many files on the Desktop of Many Monitors, Yes the data is split in many vaults(Monitors) but still it would not look cleaner and simpler to look all the the files on the desktop , yet Tags could only categorize.

Use Case: 
Consider the Following Structure

``` mermaid
flowchart  TD  
 TYPE["VPN / BANK / Cloud Provider"]  -->  pA["Provider A"]
 TYPE  -->  pB["Provider B"]
 TYPE  -->  pC["Provider C"]
pA -->  A1["Account 1"]
pA -->  A2["Account 2"]
pA -->  A3["Account 3"]

pB -->  A4["Account 4"]
pB -->  A5["Account 5"]

pC -->  A6["Account 6"]
 ```

With the current implementation, One could either 
- Have the name of the Type as a Tag, for which if he searches would result all items from all the vaults available for him.
- Create Multiple Vaults With Type And Provider (Ex: VPN - Provider A, in which all the 3 Accounts would be present, which would also require addition of members from the old vault to the new.
- Maybe something else..

Thus, I ask from you to reconsider the establishment of Folders.


I propose in the following way,
To Add additional Metadata to the Items entity, 
- folderId
- isFolder, folderPath

Flow:
Folder Creation
- Create an Item -> Select Item Type as Folder -> Enter Name
- The Folder Item will have additional metadata (isFolder:true, folderLocation:'/path/inside/vault')

Display
- if item has folderId, then show only if it is inside that folder
- if is folder, show as folder, and in case of  is sub-folder only show if inside foldeId (where folderId is parent folderId)

This way, this only masks the visibility of items without changing any crypto stature I guess. 


How does this sound? @MaKleSoft

**Comment by MaKleSoft** [comment]
@mdowais I'm still not convinced, but you're proposal is so detailed and well-argued that I've decided to reopen this issue as "under consideration". Not promising anything though 😉

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@mdowais, I fail to ascertain how organization is not better provided by tags than hierarchy, which you propose, and your statement that categorization and organization are somehow not synonymous is surely incorrect.

**Comment by mdowais** [comment]
> @mdowais I'm still not convinced, but you're proposal is so detailed and well-argued that I've decided to reopen this issue as "under consideration". Not promising anything though 😉

Thanks for the reply, hoping for the best.

**Comment by mdowais** [comment]
> @mdowais, I fail to ascertain how organization is not better provided by tags than hierarchy, which you propose, and your statement that categorization and organization are somehow not synonymous is surely incorrect.
> 
> However, one potential method of remediation of this problem is duplication of how niche tag-only filesystems support current OSes, which obviously are dependent upon hierarchical directories to organize their components. This is by assigning directories tags that are transparently interpreted as directories. I believe that this is the sole viable method of implementation because all else would introduce significant unnecessary complexity, but even this may be unnecessary.
> 
> Are you able to explain why what you propose is actually superior to tagging? My knowledge of this debate was that tags had been unanimously considered to be superior to directories since approximately 2005.

Being from Windows and Linux background,  folders create a parent-child relationship - simple, sleek.

Tags as you say have a many-to-many relationship, by which it would help finding the cousins but not the parent.

And, as in every code base, the files are inside particular folders for a reason, you can have all files inside the same folder (where-ever possible) yet, for the sake of understanding, isolating, differentiating, directories and sub-directories are created.

Tags are a beautiful thing, but they are mere Categorizing Agents, Like That is an animal, a flower, a human, etc.. (Entertainment, Bank, VPN,etc..)

But Folders are hierarchical, and offer better drill-drown approach, with superior classification, isolation.

Without folders, I fail to understand how to establish *ONE* vault, which has 5 Bank Accounts in each of the3 Banks, without folder 15 Items would be displayed on the vault, where as with folders Each Banks can have its own folder, and the accounts can be inside them.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Certainly, for the sake of security, it might be useful. I suppose that it would reduce erroneous modification of alternatively apparently duplicate entries, and ultimately it does not impact those that do not desire this feature. I am apathetic toward it.

**Comment by MaKleSoft** [comment]
@mdowais You make excellent points, and I can see now how folders could in fact add value not as alternative to tags, but as a complementary feature. I also like the implementation approach you've suggested, although I'd change a couple of minor details. I'd also have to put a little more thought into how this could be added without introducing too much clutter, and consider some technical challenges that might come with it. But I have to say I'm warming up to the idea!

**Comment by MaKleSoft** [comment]
@BrunoBernardino Do you have any thoughts on this?

**Comment by BrunoBernardino** [comment]
I understand and appreciate the value and differences between folders and tags, and the suggestion from @mdowais does seem to make it easier to not add too much complexity, other than another visual "level depth" indicator and navigation, which we currently don't _really_ have. Search would also have to take this in consideration, as people would likely want to search for folders, maybe even tag them later.

That being said, I've never found this need on password managers (can't even remember if LastPass or 1Password have something like this, for example? Maybe they do and I've never used it), as I've always had multiple vaults for "personal", "family", "company", "legal", etc. kind of separation, so I guess one could argue I've always used vaults like folders, and I don't see how this wouldn't work for @mdowais except for the lack of depth > 1.

I think this would make sense to implement given enough people asking for it, because of the added UI/UX complexity.

**Comment by mdowais** [comment]
> 

Thanks for the reply. Folders are available in [Bitwarden](https://bitwarden.com/help/folders/), [LastPass](https://support.lastpass.com/help/organize-your-vault-with-folders-lp040010).

I do understand Introducing Sub-Vault would add Complexity, so Folders would be better.

As per my understanding of the codebase, every entry inside the Vault is a `VaultItem` , And tags are also a meta data to it.

As discussed about my implementation earlier, addition of meta data to the `VaultItem` would be the easiest. 

- Let a folder be a VaultItem.
- And for the UI/UX Part, as it is an item, the same implementation would also suffice.
- if(currentItem.folderId) currentView.entries.push(...vaultItems.filter(x=>x.parentFolderId === currentItem.folderId)), and also add a first entry to go back (navigation.goback()), and browser back button will obviously handle it naturally.

**Comment by mdowais** [comment]
> I've always had multiple vaults for "personal", "family", "company", "legal", etc. kind of separation

Vaults are a separate thing, it has the ability to to maintain a ACL and crypto stature. Folders can have sub-folders.

Using Different Vaults for Family, Company, Legal is Absolutely Perfect.

But To Create N Vaults for the sake of Classification alone, would be bad. 

Example : 

With Folders (Total Vaults Needed : 1)
```
Vault: Company
- Entry(A)
- Entry(B)
- Bank (folder)
   - A Bank (Sub Folder)
         - Account A (Entry)
         - Account B (Entry)
   - B Bank (Sub-Folder)
          - Account C (Entry)
         - Account D (Entry)
```

Without Folders (Total Vaults needed : 3)
```
Vault: Company
- Entry(A)
- Entry(B)
Vault: Company-Bank-A
         - Account A (Entry)
         - Account B (Entry)
Vault: Company-Bank-B
          - Account C (Entry)
         - Account D (Entry)
```

And more the vaults -> more Processing, more storage(maybe), etc..

So, Vaults have its own purpose, Tags have their own, and Folders will have their own.

**Comment by BrunoBernardino** [comment]
@mdowais thanks for the extra explanation. Just to be clear, I wasn't saying you could replace folders with vaults, just why it was never necessary to me.

And to clarify further, the complexity we're talking about here (that we want to avoid unless enough people ask for) is mostly on the UI/UX, not the database/crypto setup, given your suggestion.

I hope that makes sense!
